core/standalone/src/main/scala/org/apache/openwhisk/standalone/KafkaLauncher.scala

```diff
@@ -0,0 +1,120 @@
+                    workDir: File,
+                    kafkaUi: Boolean)(implicit logging: Logging,
+                                      ec: ExecutionContext,
+                                      actorSystem: ActorSystem,
+                                      materializer: ActorMaterializer,
+                                      tid: TransactionId) {
+
+  def run(): Future[Seq[ServiceContainer]] = {
+    for {
+      kafkaSvcs <- runKafka()
+      uiSvcs <- if (kafkaUi) runKafkaUI() else Future.successful(Seq.empty[ServiceContainer])
+    } yield kafkaSvcs ++ uiSvcs
+  }
+
+  def runKafka(): Future[Seq[ServiceContainer]] = {
+
+    //Below setting based on https://rmoff.net/2018/08/02/kafka-listeners-explained/
+    // We configure two listeners where one is used for host based application and other is used for docker based application
+    // to connect to Kafka server running on host
+    val brokerProps = Map(
+      KafkaConfig.ListenersProp -> s"LISTENER_LOCAL://localhost:$kafkaPort,LISTENER_DOCKER://localhost:$kafkaDockerPort",
+      KafkaConfig.AdvertisedListenersProp -> s"LISTENER_LOCAL://localhost:$kafkaPort,LISTENER_DOCKER://${StandaloneDockerSupport
+        .getLocalHostIp()}:$kafkaDockerPort",
+      KafkaConfig.ListenerSecurityProtocolMapProp -> "LISTENER_LOCAL:PLAINTEXT,LISTENER_DOCKER:PLAINTEXT",
+      KafkaConfig.InterBrokerListenerNameProp -> "LISTENER_LOCAL")
+    implicit val config: EmbeddedKafkaConfig =
+      EmbeddedKafkaConfig(kafkaPort = kafkaPort, zooKeeperPort = zkPort, customBrokerProperties = brokerProps)
+
+    val t = Try {
+      EmbeddedKafka.startZooKeeper(createDir("zookeeper"))
+      EmbeddedKafka.startKafka(createDir("kafka"))
+    }
+
+    Future
+      .fromTry(t)
+      .map(
+        _ =>
+          Seq(
+            ServiceContainer(kafkaPort, s"localhost:$kafkaPort", "kafka"),
+            ServiceContainer(
+              kafkaDockerPort,
+              s"${StandaloneDockerSupport.getLocalHostIp()}:$kafkaDockerPort",
+              "kafka-docker"),
+            ServiceContainer(zkPort, "Zookeeper", "zookeeper")))
+  }
+
+  def runKafkaUI(): Future[Seq[ServiceContainer]] = {
+    val hostIp = StandaloneDockerSupport.getLocalHostIp()
+    val port = checkOrAllocatePort(9000)
+    val env = Map(
+      "ZOOKEEPER_CONNECT" -> s"$hostIp:$zkPort",
+      "KAFKA_BROKERCONNECT" -> s"$hostIp:$kafkaDockerPort",
+      "JVM_OPTS" -> "-Xms32M -Xmx64M",
+      "SERVER_SERVLET_CONTEXTPATH" -> "/")
+
+    logging.info(this, s"Starting Kafka Drop UI port: $port")
+    val name = containerName("kafka-drop-ui")
+    val params = Map("-p" -> Set(s"$port:9000"))
+    val args = createRunCmd(name, env, params)
+
+    val f = docker.runDetached("obsidiandynamics/kafdrop", args, true)
+    f.map(_ => Seq(ServiceContainer(port, s"http://localhost:$port", name)))
+  }
+
+  private def createDir(name: String) = {
+    val dir = new File(workDir, name)
+    FileUtils.forceMkdir(dir)
+    Directory(dir)
+  }
```

Comment:
Is embedded Kafka writing topic message logs to this work directory? If yes, do we clean up this data when shutting the standalone openwhisk down? If not, there is a chance that this directory grows large depending on Kafka retention policy and number of activations... If this data is only cleaned up on system reboot, the directory can grow large over time.

Reply:
Currently the dir would be cleaned upon standalone jar restart (not system restart). For now not deleting the files upon exit